src/Embedders/Word2Vec.php

```diff
@@ -0,0 +1,937 @@
+                $preppedSentence = $this->prepSentence($sentence);
+
+                $embeddings = [];
+                foreach ($preppedSentence as $word) {
+                    $embeddings[] = $this->embedWord($word);
+                }
+
+                $embeddedDataset[][$i2] = Matrix::stack($embeddings)->transpose()->mean();
+            }
+        }
+
+        return $embeddedDataset;
+    }
+
+    /**
+     * Determining the top N similar words provided an array of positive words and negative words.
+     *
+     * @param string[] $positive
+     * @param string[] $negative
+     * @param int $top
+     * @return string[] $result
+     */
+    public function most_similar(array $positive, array $negative = [], $top = 20) : array
+    {
+        $positive_array = $negative_array = $all_words = $means = [];
+
+        foreach ($positive as $word) {
+            $positive_array[$word] = 1.0;
+        }
+        
+        foreach ($negative as $word) {
+            $negative_array[$word] = -1.0;
+        }
+
+        $word_array = array_merge($positive_array, $negative_array);
+
+        foreach ($word_array as $word => $weight) {
+            $wordEmbedding = $this->wordVec($word);
+
+            if ($wordEmbedding instanceof Vector) {
+                $means[] = $wordEmbedding->multiplyScalar($weight);
+                $all_words[] = $this->vocab[$word]['index'];
+            }
+        }
+
+        if (empty($all_words)) {
+            throw new InvalidArgumentException('Positive words were not found in vocab.');
+        }
+
+        $mean = Matrix::stack($means)->transpose()->mean();
+        $l2 = Matrix::stack($this->vectors_norm);
+        $dists = $mean->transpose()->matmul($l2->transpose())->asArray()[0];
+
+        arsort($dists, 1);
+
+        $result = [];
+        foreach ($dists as $index => $weight) {
+            if (!in_array($index, $all_words)) {
+                $result[$this->index2word[$index]] = $weight;
+            }
+        }
+
+        $result = array_slice($result, 0, $top, true);
+
+        return $result;
+    }
+
+    /**
+     * Returns the word embedding for a given word.
+     *
+     * @param string $word
+     * @param bool $use_norm
+     * @return \Tensor\Vector|null $result
+     */
+    public function wordVec(string $word, bool $use_norm = true) : ?Vector
+    {
+        if (!array_key_exists($word, $this->vocab)) {
+            return null;
+        }
+
+        if ($use_norm) {
+            $result = $this->vectors_norm[$this->vocab[$word]['index']];
+        } else {
+            $result = $this->vectors[$this->vocab[$word]['index']];
+        }
+
+        return $result;
+    }
+
+    /**
+     * Returns the word embedding, or a vector of zeros if empty, for a given word.
+     *
+     * @param string $word
+     * @param bool $use_norm
+     * @return \Tensor\Vector $result
+     */
+    public function embedWord(string $word, bool $use_norm = true) : Vector
```

Comment:
After thinking about it, this method should probably return just a plain PHP array instead of a Vector object. It's leaking an implementation detail

Reply:
This is a situation where I'd have to use range and create an array of zeros, are you ok with this?

Reply:
I'm not sure how that applies here. Can you explain?

Reply:
I can't remember the specific situation regarding this one tbh.

It currently returns a vector of zeros if empty.

This is to work in tandem with returning the word vector if word exists in vocab.

Reply:
I remembered, Vector::zeros($this->dimensions) causes this fatal error. 

Fatal error: Uncaught InvalidArgumentException: The number of elements must be greater than 0, 0 given. in C:\xampp\htdocs\rubix\RubixML\vendor\rubix\tensor\src\Vector.php:72 Stack trace: #0 C:\xampp\htdocs\rubix\Extras\src\Embedders\Word2Vec.php(605): Tensor\Vector::zeros(0, 100) #1 C:\xampp\htdocs\rubix\Extras\src\Embedders\Word2Vec.php(320): Rubix\ML\Embedders\Word2Vec->prepareWeights() #2 C:\xampp\htdocs\rubix\test.php(359): Rubix\ML\Embedders\Word2Vec->train(Object(Rubix\ML\Datasets\Unlabeled)) #3 {main} thrown in C:\xampp\htdocs\rubix\RubixML\vendor\rubix\tensor\src\Vector.php on line 72